### ISSUE TITLE
[Bug] release opens apk file in beta

### ISSUE THREAD

**Issue by Djfe** [mozilla-mobile/fenix#18066]
[comment]: # (Please do your best to search for duplicate issues before filing a new issue so we can keep our issue board clean)
[comment]: # (Every issue should have the exact bug and steps to reproduce described in it. Please do not file feedback list tickets as it is difficult to parse them and address their individual points)
[comment]: # (Read https://github.com/mozilla-mobile/fenix#i-want-to-file-an-issue for more information)

## Steps to reproduce
Download apk file and open it.
https://github.com/k9mail/k-9/releases
![Screenshot_20210218-212242](https://user-images.githubusercontent.com/5685512/108418507-a6109280-7231-11eb-9557-3aa237196d7f.jpg)
![Screenshot_20210218-212249](https://user-images.githubusercontent.com/5685512/108418511-a6a92900-7231-11eb-85ea-de0399ae1b47.jpg)

### Expected behavior
no security dialogue since I already allowed the Firefox release to install apps.

### Actual behavior
It shows a security dialogue for Firefox Beta.
Why does Firefox Release open/install apk files with Firefox Beta when it's available?
Is this maybe a bug in the build system?
![Screenshot_20210218-212254](https://user-images.githubusercontent.com/5685512/108418554-b4f74500-7231-11eb-8068-12d38c91bfcf.jpg)

Edit: This never happened, when I had installed Release and Nightly in parallel. Only recently I installed Firefox Beta. This was the first time I downloaded an apk since then. (I get release, beta and Nightly through the playstore channel).

### Device information

* Device vendor / model and Android version: OnePlus 3T Android 10 Havoc OS
* Firefox for Android version: 85.1.3 and 86.0.0 beta 5

**Comment by amedyne** [comment]
Need help from QA to reproduce this.

**Comment by LaurentiuApahideanSV** [comment]
I have managed to reproduce the issue by performing the following steps:
1. Install Firefox Release (86.1.1) from the Play Store.
2. Download and install the apk from https://github.com/k9mail/k-9/releases by allowing all necessary permissions.
3. Install Firefox Preview Beta (86.0.0-beta.5) from the play Store and launch it.
4. Go back to Firefox Release and repeat step 2.

Expected behavior
The apk installs without issues.

Actual behavior
Permission to install the apk is required from Beta even if Release permission is granted.

### Device used:
- OnePlus 6T (Android 9)

**Comment by Mugurell** [comment]
This is interesting..

**Comment by Mugurell** [comment]
Can reproduce this on a Samsung S7 with Android 8 but not on a Pixel 3 with Android 11.
May be because of the release and beta using the same `sharedUserId`.

**Comment by Mugurell** [comment]
Strangely there is the exact opposite on Fennec (Release - Beta) (testing on the S7):
- opening the apk from release with beta installed doesn't show the install popup from beta
- apk can be installed from beta although in system settings I see it not having this permission

**Comment by Djfe** [comment]
I've looked a bit into what the sharedUserId is for and it doesn't seem like that has ever been the case for Firefox.
I'm wondering why sharedUserId was ever set in the AndroidManifest file.

Firefox Focus and Firefox Lockwise don't use it either.
I could understand it if Lockwise and Firefox had the same ID but not Release and Beta since they never have to talk to each other.

Regardless Google already deprecated this feature in Android 10 (API Level 29) for this exact reason: non-deterministic behavior
https://developer.android.com/guide/topics/manifest/manifest-element#uid

> **This constant was deprecated in API level 29.**
> Shared user IDs cause non-deterministic behavior within the package manager. As such, its use is strongly discouraged and may be removed in a future version of Android. Instead, apps should use proper communication mechanisms, such as services and content providers, to facilitate interoperability between shared components. Note that existing apps cannot remove this value, as migrating off a shared user ID is not supported.

**Comment by Mugurell** [comment]
I don't see anything immediately wrong or any way to improve our current implementation.
Fenix had to use the same `sharedUserId` for properly updating the old apps which were using this `sharedUserId`.

There is a bit of context in [app's `build.gradle`](https://github.com/mozilla-mobile/fenix/blob/master/app/build.gradle#L127-L133)

    // This release type is meant to replace Firefox (Release channel) and therefore needs to inherit
    // its sharedUserId for all eternity. See:
    // https://searchfox.org/mozilla-central/search?q=moz_android_shared_id&case=false&regexp=false&path=
    // Shipping an app update without sharedUserId can have
    // fatal consequences. For example see:
    //  - https://issuetracker.google.com/issues/36924841
    //  - https://issuetracker.google.com/issues/36905922


I think indeed this issue might just be something that we have to live with (similar to https://github.com/mozilla-mobile/fenix/issues/18192) and this behaviror seems to tie with Google's explanation for deprecating this API.

Also adding @pocmo who might know more.

**Comment by pocmo** [comment]
Yep, unfortunately something we have to live with. It was one of the reasons I hoped we could ship the new Firefox as a separate app. Back in Fennec we saw some bugs caused by this in situations where Android did lookups based on the Linux user ID instead of the package name. Since apps with the same sharedUserId (and certificate) share a user ID it is pretty random which app gets used internally when checking certain permissions.

**Comment by Mugurell** [comment]
Then I think we can close this as WONT_FIX. Because we really can't.
The only "solution" is to have new apps without this `sharedUserId` that users would have to fresh install.
